### PR TITLE
[Core] Adding cpp test for the builder and solvers

### PR DESCRIPTION
**Pull request by loumalouomega**
It uses the truss element from the StructuralMechanicsApplication. This test builds the truss structure created with the interface by default. Checks the block builder and solver (with/without constraints)  and the elimination builder and solver

Additionally it cleans up the builder and solver and the elimination bilder and solver

### CODE REVIEW

**Comment by loumalouomega** [comment]
I merged with #3226 to avoid conflict and compilation error

**Review by RiccardoRossi — CHANGES_REQUESTED** [review]
this is definitely not a FastPR. 

you are adding explicit to the constructor of Builder and solver for example

**Comment by loumalouomega** [comment]
I thought that the idea is to have explicit in all constructors, do I remove them?

**Comment by loumalouomega** [comment]
Which are the changes requested?

**Comment by RiccardoRossi** [comment]
adding explicit should not go in a PR called "Adding cpp test for the builder and solvers" even less so if you mark it as "FastPR"

**Review by RiccardoRossi — COMMENTED** [review]
you changed a lot of things in this PR. They are NOT cosmetic and they will take time to review

**Comment by loumalouomega** [comment]
> I merged with #3226 to avoid conflict and compilation error

I said that #3226 should go first

And the other changes are sthetic

**Comment by loumalouomega** [comment]
Now only tests are changed

**Comment by loumalouomega** [comment]
@RiccardoRossi Now only contains the tests

**Comment by loumalouomega** [comment]
@KratosMultiphysics/technical-committee @RiccardoRossi  This is just one file adding a test

File: kratos/tests/strategies/builder_and_solvers/test_builder_and_solver.cpp [hunk 1]
```diff
@@ -0,0 +1,841 @@
...
+            KRATOS_CHECK_LESS_EQUAL(std::abs((rA(8,8) - 1)/rA(8,8)), tolerance);
+        }
+
+        /**
+         * Checks if the elimination builder and solver performs correctly the resolution of the system
+         */
+        KRATOS_TEST_CASE_IN_SUITE(BasicDisplacementEliminationBuilderAndSolver, KratosCoreFastSuite)
+        {
+            if (!KratosComponents<Element>::Has("TrussElement3D2N")) {
+                std::cout << "Please compile the StructuralMechanicsApplication in order to run this test" << std::endl;
```
Comment (RiccardoRossi) [inline kratos/tests/strategies/builder_and_solvers/test_builder_and_solver.cpp]: sholdn't this be a warning? also, is this the "official" way of skipping a test?

**Review by RiccardoRossi — COMMENTED** [review]
the bad thing of this is that it requires the StructuralMechanicsApplication while testing the core.

did u check if it runs correctly if the app is not compiled? 

also, wouldn't it be better to define a "dummy_test_element" in the core and use that?

File: kratos/tests/strategies/builder_and_solvers/test_builder_and_solver.cpp (hunk 1 above)
Comment (loumalouomega) [reply kratos/tests/strategies/builder_and_solvers/test_builder_and_solver.cpp]: This is the way I has been using in my tests for mixed applications. And yes it works without compiling the Structural

**Comment by loumalouomega** [comment]
Your request is to add the dummy_bar_element?

**Comment by loumalouomega** [comment]
I created a dummy bar element to not depend on StructuralMechanicsApplication

File: kratos/tests/strategies/builder_and_solvers/test_builder_and_solver.cpp [hunk 2]
```diff
@@ -0,0 +1,835 @@
...
+            LinearSolverType::Pointer p_solver = LinearSolverType::Pointer( new SkylineLUFactorizationSolverType() );
+            BuilderAndSolverType::Pointer p_builder_and_solver = BuilderAndSolverType::Pointer( new ResidualBasedBlockBuilderAndSolverWithConstraintsType(p_solver) );
+
+            const SparseSpaceType::MatrixType& rA = BuildSystem(r_model_part, p_scheme, p_builder_and_solver);
+
+            // To create the solution of reference
+//             DebugLHS(rA);
+
+            // The solution check
+            constexpr double tolerance = 1e-4;
```
Comment (RiccardoRossi) [inline kratos/tests/strategies/builder_and_solvers/test_builder_and_solver.cpp]: much stricter tolerance!!

File: kratos/tests/strategies/builder_and_solvers/test_builder_and_solver.cpp [hunk 3]
```diff
@@ -0,0 +1,835 @@
...
+            KRATOS_CHECK_LESS_EQUAL(std::abs((rA(20,20) - 1.77473e+09)/rA(20,20)), tolerance);
+            KRATOS_CHECK_LESS_EQUAL(std::abs((rA(20,21) - -3.70114e+08)/rA(20,21)), tolerance);
+            KRATOS_CHECK_LESS_EQUAL(std::abs((rA(21,16) - 3.70114e+08)/rA(21,16)), tolerance);
+            KRATOS_CHECK_LESS_EQUAL(std::abs((rA(21,17) - -1.85057e+08)/rA(21,17)), tolerance);
+            KRATOS_CHECK_LESS_EQUAL(std::abs((rA(21,20) - -3.70114e+08)/rA(21,20)), tolerance);
+            KRATOS_CHECK_LESS_EQUAL(std::abs((rA(21,21) - 1.85057e+08)/rA(21,21)), tolerance);
+        }
+
+        /**
+         * Checks if the block builder and solver with constraints performs correctly the resolution of the system
```
Comment (RiccardoRossi) [inline kratos/tests/strategies/builder_and_solvers/test_builder_and_solver.cpp]: the assembel, not the solution

**Comment by loumalouomega** [comment]
@RiccardoRossi can we merge this?

**Comment by pooyan-dadvand** [comment]
I wouldn't go for a dummy element bar. We have some elements in core that you can use. (i.e. DistanceCalculationElement)

**Comment by loumalouomega** [comment]
@pooyan-dadvand I used the bar from the StructuralMechanicsApplication because we can do the connectivity by hand (quite common in structural analysis). @RiccardoRossi told me to create a dummy element, I think that using the distance calculation element breaks the meaning of the test

**Comment by loumalouomega** [comment]
@pooyan-dadvand The element is only compiled if the tests are compiled (it is added directly in the tests folder), it is not registered and not accessible that from a cpp test

**Comment by pooyan-dadvand** [comment]
As I mentioned in other PR this the bad point about this Mock up element that is not a real one and may become a source of false ok in tests. Meanwhile checking with the distance one can make a better coverage.

**Comment by loumalouomega** [comment]
@RiccardoRossi, @pooyan-dadvand has mentioned me that we can proceed with the mock up element

**Comment by loumalouomega** [comment]
@RiccardoRossi

**Review by RiccardoRossi — APPROVED** [review]

File: kratos/tests/strategies/builder_and_solvers/test_builder_and_solver.cpp (hunk 2 above)
Comment (loumalouomega) [reply kratos/tests/strategies/builder_and_solvers/test_builder_and_solver.cpp]: @RiccardoRossi This is outdated

imagen

File: kratos/tests/strategies/builder_and_solvers/test_builder_and_solver.cpp (hunk 3 above)
Comment (loumalouomega) [reply kratos/tests/strategies/builder_and_solvers/test_builder_and_solver.cpp]: Idem 

imagen

**Comment by loumalouomega** [comment]
@RiccardoRossi for some reason GitHub is showing old pieces of code

**Pull request merged**